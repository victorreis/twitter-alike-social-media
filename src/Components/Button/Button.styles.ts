import styled, { CSSObject, DefaultTheme } from 'styled-components';

import { FeedbackColor } from '../../Theme/Types/Colors.types';
import { toPx } from '../../Utils/Transform/toPx.util';
import { ButtonSize, ButtonSizes, ButtonStyleProps } from './Button.types';

const getBackgroundColor = ({
  theme,
  hoverFeedbackColor,
  hover,
  disabled,
}: {
  theme: DefaultTheme;
  hoverFeedbackColor?: FeedbackColor;
  hover: boolean;
  disabled: boolean;
}): string => {
  if (disabled) {
    return theme.colors.background.default.dark;
  }
  if (hover && hoverFeedbackColor) {
    return theme.colors.feedback[hoverFeedbackColor].dark;
  }
  return theme.colors.background.default.darkest;
};

const getButtonDimensions = ({ size }: { size: ButtonSize }): CSSObject => {
  const buttonDimensions: ButtonSizes = {
    MD: '2.2rem',
    LG: '3rem',
  };
  return { height: buttonDimensions[size] };
};

const getBorderColor = ({
  theme,
  hoverFeedbackColor,
  hover,
  disabled,
}: {
  theme: DefaultTheme;
  hoverFeedbackColor?: FeedbackColor;
  hover: boolean;
  disabled: boolean;
}): string => {
  if (hover && hoverFeedbackColor) {
    if (disabled) {
      return theme.colors.feedback[hoverFeedbackColor].dark;
    }
    return theme.colors.feedback[hoverFeedbackColor].darkest;
  }
  if (disabled) {
    return `${theme.colors.main.effect.normal}33`;
  }
  return `${theme.colors.main.effect.normal}55`;
};

export const ButtonContainer = styled.button<ButtonStyleProps>`
  cursor: pointer;
  border: 0;
  letter-spacing: 0.2rem;
  text-transform: uppercase;
  align-items: center;
  padding-left: 1rem;
  padding-right: 1rem;
  box-sizing: border-box;

  ${getButtonDimensions};

  border-width: 1px;
  border-style: solid;
  border-color: ${({ theme, hoverFeedbackColor, disabled }) =>
    getBorderColor({ theme, hoverFeedbackColor, hover: false, disabled })};
  border-radius: ${({ theme }) => toPx(theme.borders.radius.LG)};

  ${({ theme }) => theme.typographies.button};
  color: ${({ theme, disabled }) =>
    disabled
      ? theme.colors.background.default.light
      : theme.colors.background.default.lightest};

  background-color: ${({ theme, disabled }) =>
    getBackgroundColor({ theme, hover: false, disabled })};

  &:hover {
    background-color: ${({ theme, hoverFeedbackColor, disabled }) =>
      getBackgroundColor({ theme, hoverFeedbackColor, hover: true, disabled })};

    border-color: ${({ theme, hoverFeedbackColor, disabled }) =>
      getBorderColor({ theme, hoverFeedbackColor, hover: true, disabled })};

    ${({ theme, hoverFeedbackColor, disabled }) =>
      !disabled &&
      `box-shadow: inset 0 0 1em ${getBorderColor({
        theme,
        hoverFeedbackColor,
        hover: true,
        disabled,
      })}`};
  }
`;
