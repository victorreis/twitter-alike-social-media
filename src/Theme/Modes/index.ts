export { darkTheme } from './Dark.theme';
