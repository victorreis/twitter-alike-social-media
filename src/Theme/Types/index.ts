export * from './Borders.types';
export * from './Colors.types';
export * from './CustomTheme.types';
export * from './Sizes.types';
export * from './Typographies.types';
