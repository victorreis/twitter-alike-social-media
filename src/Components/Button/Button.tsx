import { useState } from 'react';

import { TestProps } from '../../Config/Tests/Test.types';
import { ButtonContainer } from './Button.styles';
import { ButtonProps, DefaultButtonProps } from './Button.types';

export const buttonDefaults: Required<DefaultButtonProps> &
  Required<TestProps> = {
  testID: 'Button',
  type: 'button',
  size: 'LG',
  disabled: false,
};

export const Button: React.FC<ButtonProps> = (props): JSX.Element => {
  const {
    testID = buttonDefaults.testID,
    children,
    type = buttonDefaults.type,
    size = buttonDefaults.size,
    disabled = buttonDefaults.disabled,
    onClick,
    hoverFeedbackColor,
    hoverText,
    ...others
  } = props;

  const [text, setText] = useState<string>(children);

  const handleClick = (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    event.stopPropagation();
    if (onClick) {
      onClick();
    }
  };

  const handleMouseEnter = () => {
    if (hoverText) {
      setText(hoverText);
    }
  };
  const handleMouseLeave = () => {
    setText(children);
  };

  return (
    <ButtonContainer
      data-testid={testID}
      disabled={disabled}
      hoverFeedbackColor={hoverFeedbackColor}
      hoverText={hoverText}
      onClick={handleClick}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      size={size}
      type={type}
      {...others}
    >
      {text}
    </ButtonContainer>
  );
};
