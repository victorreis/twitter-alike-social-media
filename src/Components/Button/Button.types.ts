import { TestProps } from '../../Config/Tests/Test.types';
import { Size } from '../../Theme/Types';
import { FeedbackColor } from '../../Theme/Types/Colors.types';

export const availableButtonTypes = ['button', 'submit'] as const;
export type ButtonType = typeof availableButtonTypes[number];

export const availableButtonSizes: ReadonlyArray<Extract<Size, 'MD' | 'LG'>> = [
  'MD',
  'LG',
] as const;
export type ButtonSize = typeof availableButtonSizes[number];
export type ButtonSizes = Record<ButtonSize, string>;

export interface RequiredButtonProps {
  /**
   * Component's children.
   */
  children: string;

  /**
   * Callback function that is called when the component is clicked.
   */
  onClick: () => void;
}

export interface DefaultButtonProps {
  /**
   * Sets the button type.
   * @default 'button'.
   */
  type?: ButtonType;

  /**
   * Sets the button size.
   * @default 'LG'.
   */
  size?: ButtonSize;

  /**
   * Disable the button, so it can't be clicked.
   * @default 'false'.
   */
  disabled?: boolean;
}

export interface OptionalButtonProps {
  /**
   * Sets the background color when the component is in a hover state.
   */
  hoverFeedbackColor?: FeedbackColor;

  /**
   * Sets the button text when the component is in a hover state.
   */
  hoverText?: string;
}

export type ButtonProps = RequiredButtonProps &
  DefaultButtonProps &
  OptionalButtonProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLButtonElement>, 'children' | 'onClick'>;

export type ButtonStyleProps = Required<DefaultButtonProps> & {
  hoverFeedbackColor?: FeedbackColor;
  hoverText?: string;
};
