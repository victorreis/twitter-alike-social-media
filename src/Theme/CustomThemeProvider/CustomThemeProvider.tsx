/* eslint-disable @typescript-eslint/ban-ts-comment */
import { useMemo, useState, createContext } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'styled-components';

import { darkTheme } from '../Modes';
import { ThemeMode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
  DefaultCustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<ThemeMode | 'default', CustomThemeType> = {
  default: darkTheme,
  dark: darkTheme,
};

export const ThemeContext = createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<DefaultCustomThemeProviderProps> =
  {
    testID: 'ThemeContextProvider',
    themeName: 'dark',
  };

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: ThemeMode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      {/* waiting for a new lib version compatible with React 18 */}
      {/* @ts-ignore */}
      <ThemeProvider theme={theme}>
        <BrowserRouter>{children}</BrowserRouter>
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
