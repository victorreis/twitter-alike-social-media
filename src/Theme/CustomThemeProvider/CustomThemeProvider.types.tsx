import { CustomThemeType, ThemeMode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: ThemeMode) => void;
};

export interface RequiredCustomThemeProviderProps {
  children?: React.ReactNode;
}

export interface DefaultCustomThemeProviderProps {
  testID?: string;
  themeName?: ThemeMode;
}

export type CustomThemeProviderProps = RequiredCustomThemeProviderProps &
  DefaultCustomThemeProviderProps;
