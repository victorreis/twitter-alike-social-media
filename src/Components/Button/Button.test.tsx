import {
  fireEvent,
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { hexToRgb } from '../../Utils/Transform/hexToRgb.util';
import { toPx } from '../../Utils/Transform/toPx.util';
import { Button, buttonDefaults } from './Button';
import { RequiredButtonProps, ButtonProps } from './Button.types';

describe('button component tests', () => {
  const text = 'text';
  const newType = 'submit';
  const onClick = jest.fn();

  const requiredProps: RequiredButtonProps = {
    children: text,
    onClick,
  };

  const setup = (props?: ButtonProps) => {
    const renderRTR = () =>
      renderRTRCreator<ButtonProps>(Button, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<ButtonProps>(Button, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it('should render the component', () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(buttonDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it('should render the text', () => {
      expect.assertions(1);
      setup().renderJestDom();
      const element = screen.getByText(text);

      expect(element).toBeInTheDocument();
    });

    it(`should call the 'onCLick' callback when the button is clicked`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(buttonDefaults.testID);

      fireEvent.click(testInstance);

      expect(onClick).toHaveBeenCalledWith();
    });

    it(`should render '${buttonDefaults.type}' as the default 'type'`, () => {
      expect.assertions(1);
      const instance = setup().renderRTR().root;
      const element = instance.findByProps({
        type: buttonDefaults.type,
      });

      expect(element).toBeTruthy();
    });

    it(`should override the default 'type' when it is passed as prop`, () => {
      expect.assertions(1);
      const instance = setup({
        ...requiredProps,
        type: newType,
      }).renderRTR().root;
      const element = instance.findByProps({ type: newType });

      expect(element).toBeTruthy();
    });
  });

  describe('style tests', () => {
    it('should have styled the correct styles', () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(buttonDefaults.testID);

      expect(container).toHaveStyle({
        backgroundColor: hexToRgb(
          themes.default.colors.background.default.darkest
        ),
        borderRadius: toPx(themes.default.borders.radius.LG),
        color: hexToRgb(themes.default.colors.background.default.lightest),
        ...themes.default.typographies.button,
      });
    });
  });

  describe('snapshot tests', () => {
    it('should render correctly', () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
