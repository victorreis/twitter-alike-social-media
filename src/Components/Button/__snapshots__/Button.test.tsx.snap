// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button component tests snapshot tests should render correctly 1`] = `
<button
  className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 iHxNIv"
  data-testid="Button"
  disabled={false}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  size="LG"
  type="button"
>
  text
</button>
`;
